mobile/android/genymotion/genymotion-service.ts

```diff
@@ -0,0 +1,152 @@
+import { AndroidVirtualDevice, DeviceTypes, NOT_RUNNING_EMULATOR_STATUS } from "../../../constants";
+import { EOL } from "os";
+
+export class AndroidGenymotionService implements Mobile.IAndroidVirtualDeviceService {
+	constructor(private $adb: Mobile.IAndroidDebugBridge,
+		private $childProcess: IChildProcess,
+		private $devicePlatformsConstants: Mobile.IDevicePlatformsConstants,
+		private $emulatorHelper: Mobile.IEmulatorHelper,
+		private $fs: IFileSystem,
+		private $hostInfo: IHostInfo,
+		private $logger: ILogger,
+		private $virtualBoxService: Mobile.IVirtualBoxService) { }
+
+	public async getAvailableEmulators(adbDevicesOutput: string[]): Promise<Mobile.IAvailableEmulatorsOutput> {
+		const availableEmulatorsOutput = await this.getAvailableEmulatorsCore();
+		const genies = availableEmulatorsOutput.devices;
+		const runningEmulatorIds = await this.getRunningEmulatorIds(adbDevicesOutput);
+		const runningEmulators = await Promise.all(runningEmulatorIds.map(emulatorId => this.getRunningEmulatorData(emulatorId, genies)));
+		const devices = availableEmulatorsOutput.devices.map(emulator => this.$emulatorHelper.getEmulatorByImageIdentifier(emulator.imageIdentifier, runningEmulators) || emulator);
+		return {
+			devices,
+			errors: availableEmulatorsOutput.errors
+		};
+	}
+
+	public async getRunningEmulatorIds(adbDevicesOutput: string[]): Promise<string[]> {
+		const results = await Promise.all<string>(
+			<Promise<string>[]>(_(adbDevicesOutput)
+				.filter(r => !r.match(AndroidVirtualDevice.RUNNING_AVD_EMULATOR_REGEX))
+				.map(async row => {
+					const match = row.match(/^(.+?)\s+device$/);
+					if (match && match[1]) {
+						// possible genymotion emulator
+						const emulatorId = match[1];
+						const result = await this.isGenymotionEmulator(emulatorId) ? emulatorId : undefined;
+						return Promise.resolve(result);
+					}
+
+					return Promise.resolve(undefined);
+				}).value())
+		);
+
+		return _(results).filter(r => !!r)
+			.map(r => r.toString())
+			.value();
+	}
+
+	public startEmulator(imageIdentifier: string): void {
+		const pathToPlayer = this.$fs.exists(this.defaultPlayerPath) ? this.defaultPlayerPath : "player";
+		try {
+			return this.$childProcess.spawn(pathToPlayer, ["--vm-name", imageIdentifier], { stdio: "ignore", detached: true }).unref();
+		} catch (err) {
+			this.$logger.trace(`error while starting emulator. More info: ${err}`);
```

Comment:
this catch will never catch anything as we are using async version of the spawn. In case there's error, it will be thrown on nextTick. As we do not have `.on("error"` handler for the currently spawned process, the error will be raised as UncaughtException and could lead to many unexpected behaviors.
Also why is the process fired and forget?